src/passes/OptimizeInstructions.cpp

```diff
@@ -863,13 +863,14 @@ struct OptimizeInstructions
           // sides are identical, fold
           // if we can replace the if with one arm, and no side effects in the
           // condition, do that
-          auto needCondition = effects(iff->condition).hasSideEffects();
-          auto isSubType = Type::isSubType(iff->ifTrue->type, iff->type);
-          if (isSubType && !needCondition) {
+          bool needCondition = effects(iff->condition).hasSideEffects();
+          bool wouldBecomeUnreachable =
+            iff->type.isConcrete() && iff->ifTrue->type == Type::unreachable;
+          if (!wouldBecomeUnreachable && !needCondition) {
             return iff->ifTrue;
           } else {
             Builder builder(*getModule());
-            if (isSubType) {
+            if (!wouldBecomeUnreachable) {
               return builder.makeSequence(builder.makeDrop(iff->condition),
                                           iff->ifTrue);
             } else {
```

Comment:
the comment on the line below, could be made more specific, to replace "not" with "unreachable".

or the comment can be removed - the assertion says it all.

(I can't add a github suggestion because it's out of the range of the diff...)